_posts/2017/2017-02-20-continuous-documentation-with-ansible.md

```diff
@@ -0,0 +1,207 @@
+In the case of "management code" this is more difficult, because the things being modelled are often heterogeneous and not as focussed in terms of their implementation as for example a REST-API or a library / module.  
+The concept of [using declarative formats for automation](https://12factor.net/#introduction) can fill this gap.
+
+# Structure of Ansible projects
+As there are many sources of Ansible best practices to be found on the web, instead of writing another one, I would like to assume the viewpoint of someone reading a Ansible code, rather than writing it.
+Note that I am using the term "Ansible project" without implying a certain structure in terms of source code repositories.  
+
+## Playbooks
+[Playbooks](http://docs.ansible.com/ansible/playbooks.html) are the entry point when looking at an Ansible project.  
+A common convention is two have two kinds of Playbooks.
+
+### Setup Playbooks
+One type of playbook would initially provision / prepare an environment to deploy an application into it.  
+These playbooks would typically be named `setup-<project>.yml`.
+While this type could traditionally be seen more on the IT department's side of an organization, this type of playbook would also be used [to set up a development environment](https://github.com/ePages-de/mac-dev-setup) (i.e. developer laptop) for that application.  
+The point of using the same technique for both of these things is that it becomes easier to manage and document differences between them in a joint effort. This continually provides value for the company, because setup times for developers are reduced.  
+Using playbooks to install a development environment is a good way of incorporating new developers. Invariably occurring problems are an opportunity to familiarize with the environment and getting used to the workflow by documenting their solution.
+
+### Deployment Playbooks
+The other kind of playbooks would be named `deploy-<project>.yml`.  
+These are meant to execute continuously, deploying changing versions of an application.  
+This type of playbook would typically be executed during a CI/CD run, and a failure during that phase could be the cause for investigating the playbook in the first place.
+
+Let's look at an example playbook (`deploy-unity.yml`):
+{% highlight yml %}
+---
+- hosts: application-vms
+
+  remote_user: deploy
+  become: true
+
+  pre_tasks:
+    - name: Abort on missing ePages-installation
+      shell: rpm -q epages
+      args:
+        warn: false
+      register: check_epages
+      failed_when: "check_epages.rc != 0"
+      changed_when: false
+
+  roles:
+    - role: docker
+      tags: [ docker ]
+
+    - role: epages-unity
+      tags: [ unity ]
+{% endhighlight %}
+This playbook should be fairly self-explanatory, given you know what "[Unity](https://www.epages.com/de/now/)" is.    
```

Comment:
Please avoid mentioning "Unity" - as publicly we communicate the product as "ePages Now". So please go with a "fake" :-)